### ISSUE TITLE
"malformed" xml disrupt salt deploy (apply) for samba shares

### ISSUE THREAD

**Issue by mneiger** [openmediavault/openmediavault#778]
### Description of issue/question
"malformed" config.xml file will trigger bug in salt deploy despite xml being strictly ok xmlwise

### Steps to reproduce issue
(Include debug logs if possible and relevant.)

"empty" Privileges section for shared folders <privileges> sections in config.xml will trigger the error if it contains extraneous LF

salt expect a strictly empty section
> `<privileges></privileges>`

while 

```
> <privileges>
> </privileges>
```
will trigger the error although it is valid xml, and I believe should be accepted. I imagine it may be due to the use of a non xml focused text parser.
If the config.xml contains such a section, making changes to the SMB shares will trigger the error on "Apply" 

```
> Error #0:
> OMV\ExecException: Failed to execute command 'export PATH=/bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin; export LANG=C.UTF-8; omv-salt deploy run samba 2>&1' with exit code '1': yohanan:
> ----------
>           ID: configure_samba_global
>     Function: file.managed
>         Name: /etc/samba/smb.conf
>       Result: True
>      Comment: File /etc/samba/smb.conf is in the correct state
>      Started: 13:02:35.394362
>     Duration: 256.72 ms
>      Changes:
> ----------
>           ID: configure_samba_shares
>     Function: file.append
>         Name: /etc/samba/smb.conf
>       Result: False
>      Comment: An exception occurred in this state: Traceback (most recent call last):
>                 File "/usr/lib/python3/dist-packages/salt/utils/templates.py", line 400, in render_jinja_tmpl
>                   output = template.render(**decoded_context)
>                 File "/usr/lib/python3/dist-packages/jinja2/asyncsupport.py", line 76, in render
>                   return original_render(self, *args, **kwargs)
>                 File "/usr/lib/python3/dist-packages/jinja2/environment.py", line 1008, in render
>                   return self.environment.handle_exception(exc_info, True)
>                 File "/usr/lib/python3/dist-packages/jinja2/environment.py", line 780, in handle_exception
>                   reraise(exc_type, exc_value, tb)
>                 File "/usr/lib/python3/dist-packages/jinja2/_compat.py", line 37, in reraise
>                   raise value.with_traceback(tb)
>                 File "<template>", line 96, in top-level template code
>               jinja2.exceptions.UndefinedError: 'str object' has no attribute 'privilege'
> 
>               During handling of the above exception, another exception occurred:
> 
>               Traceback (most recent call last):
>                 File "/usr/lib/python3/dist-packages/salt/state.py", line 2154, in call
>                   *cdata["args"], **cdata["kwargs"]
>                 File "/usr/lib/python3/dist-packages/salt/loader.py", line 2087, in wrapper
>                   return f(*args, **kwargs)
>                 File "/usr/lib/python3/dist-packages/salt/states/file.py", line 6255, in append
>                   source_list=sl_, template=template, defaults=defaults, context=context
>                 File "/usr/lib/python3/dist-packages/salt/states/file.py", line 1245, in _get_template_texts
>                   source, "", template=template, saltenv=__env__, context=tmpctx, **kwargs
>                 File "/usr/lib/python3/dist-packages/salt/modules/cp.py", line 327, in get_template
>                   return _client().get_template(path, dest, template, makedirs, saltenv, **kwargs)
>                 File "/usr/lib/python3/dist-packages/salt/fileclient.py", line 807, in get_template
>                   data = salt.utils.templates.TEMPLATE_REGISTRY[template](sfn, **kwargs)
>                 File "/usr/lib/python3/dist-packages/salt/utils/templates.py", line 166, in render_tmpl
>                   output = render_str(tmplstr, context, tmplpath)
>                 File "/usr/lib/python3/dist-packages/salt/utils/templates.py", line 407, in render_jinja_tmpl
>                   raise SaltRenderError("Jinja variable {0}{1}".format(exc, out), buf=tmplstr)
>               salt.exceptions.SaltRenderError: Jinja variable 'str object' has no attribute 'privilege'
>      Started: 13:02:35.651586
>     Duration: 486.756 ms
>      Changes:
> ----------
>           ID: configure_samba_recyclebin_cron
>     Function: file.managed
>         Name: /etc/cron.daily/openmediavault-samba-recycle
>       Result: True
>      Comment: File /etc/cron.daily/openmediavault-samba-recycle is in the correct state
>      Started: 13:02:36.138842
>     Duration: 8.114 ms
>      Changes:
> ----------
>           ID: remove_samba_recyclebin_cron_scripts
>     Function: module.run
>       Result: True
>      Comment: file.find: []
>      Started: 13:02:36.150271
>     Duration: 7.255 ms
>      Changes:
>               ----------
>               file.find:
> ----------
>           ID: configure_default_wsdd
>     Function: file.managed
>         Name: /etc/default/wsdd
>       Result: True
>      Comment: File /etc/default/wsdd is in the correct state
>      Started: 13:02:36.157931
>     Duration: 7.353 ms
>      Changes:
> ----------
>           ID: test_samba_service_config
>     Function: cmd.run
>         Name: samba-tool testparm --suppress-prompt
>       Result: True
>      Comment: Command "samba-tool testparm --suppress-prompt" run
>      Started: 13:02:36.170495
>     Duration: 742.117 ms
>      Changes:
>               ----------
>               pid:
>                   2964
>               retcode:
>                   0
>               stderr:
>               stdout:
>                   # Global parameters
>                   [global]
>                   	allow insecure wide links = Yes
>                   	disable spoolss = Yes
>                   	domain master = Yes
>                   	encrypt passwords = Yes
>                   	get quota command = /opt/bin/samba-btrfs-quota.sh
>                   	guest account = nobody
>                   	log file = /var/log/samba/log.%m
>                   	logging = syslog
>                   	log level = 1
>                   	map to guest = Bad User
>                   	max log size = 1000
>                   	pam password change = Yes
>                   	panic action = /usr/share/samba/panic-action %d
>                   	passdb backend = tdbsam
>                   	passwd chat = *Enter\snew\s*\spassword:* %n\n *Retype\snew\s*\spassword:* %n\n *password\supdated\ssuccessfully* .
>                   	passwd program = /usr/bin/passwd %u
>                   	preferred master = Yes
>                   	printcap name = /dev/null
>                   	server min protocol = SMB2
>                   	server string = %h server
>                   	socket options = TCP_NODELAY IPTOS_LOWDELAY
>                   	time server = Yes
>                   	workgroup = NEIGER
>                   	idmap config * : range = 30000-40000
>                   	idmap config * : backend = tdb
>                   	fruit:aapl = yes
>                   	aio read size = 1
>                   	aio write size = 1
>                   	create mask = 0777
>                   	directory mask = 0777
>                   	use sendfile = Yes
> ----------
>           ID: start_samba_service
>     Function: service.running
>         Name: smbd
>       Result: False
>      Comment: One or more requisite failed: omv.deploy.samba.20shares.configure_samba_shares
>      Started: 13:02:36.986087
>     Duration: 0.02 ms
>      Changes:
> ----------
>           ID: start_samba_service_nmbd
>     Function: service.running
>         Name: nmbd
>       Result: False
>      Comment: One or more requisite failed: omv.deploy.samba.default.start_samba_service
>      Started: 13:02:36.987106
>     Duration: 0.014 ms
>      Changes:
> ----------
>           ID: start_wsdd_service
>     Function: service.running
>         Name: wsdd
>       Result: True
>      Comment: The service wsdd is already running
>      Started: 13:02:36.987371
>     Duration: 153.448 ms
>      Changes:
> 
> Summary for yohanan
> ------------
> Succeeded: 6 (changed=2)
> Failed:    3
> ------------
> Total states run:     9
> Total run time:   1.662 s in /usr/share/php/openmediavault/system/process.inc:182
> Stack trace:
> #0 /usr/share/php/openmediavault/engine/module/serviceabstract.inc(60): OMV\System\Process->execute()
> #1 /usr/share/openmediavault/engined/rpc/config.inc(167): OMV\Engine\Module\ServiceAbstract->deploy()
> #2 [internal function]: Engined\Rpc\Config->applyChanges(Array, Array)
> #3 /usr/share/php/openmediavault/rpc/serviceabstract.inc(123): call_user_func_array(Array, Array)
> #4 /usr/share/php/openmediavault/rpc/serviceabstract.inc(149): OMV\Rpc\ServiceAbstract->callMethod('applyChanges', Array, Array)
> #5 /usr/share/php/openmediavault/rpc/serviceabstract.inc(588): OMV\Rpc\ServiceAbstract->OMV\Rpc\{closure}('/tmp/bgstatusp0...', '/tmp/bgoutput8d...')
> #6 /usr/share/php/openmediavault/rpc/serviceabstract.inc(159): OMV\Rpc\ServiceAbstract->execBgProc(Object(Closure))
> #7 /usr/share/openmediavault/engined/rpc/config.inc(189): OMV\Rpc\ServiceAbstract->callMethodBg('applyChanges', Array, Array)
> #8 [internal function]: Engined\Rpc\Config->applyChangesBg(Array, Array)
> #9 /usr/share/php/openmediavault/rpc/serviceabstract.inc(123): call_user_func_array(Array, Array)
> #10 /usr/share/php/openmediavault/rpc/rpc.inc(86): OMV\Rpc\ServiceAbstract->callMethod('applyChangesBg', Array, Array)
> #11 /usr/sbin/omv-engined(537): OMV\Rpc\Rpc::call('Config', 'applyChangesBg', Array, Array, 1)
> #12 {main}
```

### Versions report

Openmediavault 5.5.5 (Upgraded to 5.5.6 since)
Debian 10

**Comment by votdev** [comment]
I don’t think this can be fixed, did not find an option for the lxml Python library to be more tolerant when parsing such an file.

I think this problem only occurs when the file is manually modified. In this case you need to accept that there are rules that must apply to be a valid XML file.

**Comment by mneiger** [comment]
Many thanks for checking. I actually expected such a outcome, but at least it's now documented, even if not fixable.
I don't remember fiddling with that part of the config.xml , but I would plead guilty for a few other bits. It could however be a some remnant of the OMV4 to OMV5 upgrade.

**Comment by votdev** [comment]
Please note that `<privileges></privileges>` is not equal to 
```
<privileges>
</privileges>
```
The first one is an empty XML element.

`[Definition: An element with no content is said to be empty.] The representation of an empty element is either a start-tag immediately followed by an end-tag, or an empty-element tag. [Definition: An empty-element tag takes a special form:]`, see https://www.w3.org/TR/xml/#sec-starttags.

The second example is an XML element containing a string which is a line break.

According to the [OMV datamodel](https://github.com/openmediavault/openmediavault/blob/master/deb/openmediavault/usr/share/openmediavault/datamodels/conf.system.sharedfolder.json#L28) the `privileges` element is a list of `privilege` elements. An empty string is faulty in this case.

The only thing which makes me wonder is that the error does not appear at the point where i expect it (where the loaded content is compared and evaluated to the datamodel), but i will investigate that.

**Comment by votdev** [comment]
I cannot reproduce the problem with the following database content:
```
    <shares>
      <!--
                        <sharedfolder>
                                <uuid>xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx</uuid>
                                <name>xxx</name>
                                <comment>xxx</comment>
                                <mntentref>xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx</mntentref>
                                <reldirpath>xxx</reldirpath>
                                <privileges>
                                        <privilege>
                                                <type>user|group</type>
                                                <name>xxx</name>
                                                <perms>0 = no permission|5 = read and execute|7 = read, wri$
                                        </privilege>
                                </privileges>
                        </sharedfolder>
                        -->
      <sharedfolder>
        <uuid>db30a097-e456-4fa5-b44b-de9aff5a0789</uuid>
        <name>test</name>
        <comment></comment>
        <mntentref>91aa3d0d-5a37-4a09-bef1-167df956863f</mntentref>
        <reldirpath>test/</reldirpath>
        <privileges>
        </privileges>
      </sharedfolder>
    </shares>
```

**Comment by votdev** [comment]
Please note that `<privileges></privileges>` is not equal to
```
<privileges>
</privileges>
```
The first one is an empty XML element.

[Definition: An element with no content is said to be empty.] The representation of an empty element is either a start-tag immediately followed by an end-tag, or an empty-element tag. [Definition: An empty-element tag takes a special form:], see https://www.w3.org/TR/xml/#sec-starttags.

The second example is an XML element containing a string which is a line break.

According to the OMV datamodel the privileges element is a list of privilege elements. An empty string is faulty in this case. Because of the datamodel and how XML handles empty elements i don't believe that your database file is generated by OMV or any XML tool like xmlstarlet. It must be modified manually.

The only thing which makes me wonder is that the error does not appear at the point where i expect it (where the loaded content is compared and evaluated to the datamodel), but i will investigate that.

**Comment by mneiger** [comment]
Thanks for the clarification.

Here is the shares section as it was when triggering the error, it this may help (I'll send it with uuid in private if necessary).
I admit I did modify it manually, during the OMV4-5 update, and when I reorganized my data disks. So I'm definitely responsible for introducing abnormalities in the file. 

[yohanan-shares.txt](https://github.com/openmediavault/openmediavault/files/5047546/yohanan-shares.txt)
